File: stock_platform/python/backend/python/services/statistics_service.py
from models import Trade, Account, Position
from schemas import StatisticsResponse, SummaryResponse
from typing import Optional
from datetime import datetime
from services import stock_price_service


async def get_statistics(
    account_id: Optional[int] = None,
    start_date: Optional[datetime] = None,
    end_date: Optional[datetime] = None
) -> StatisticsResponse:
    # 从持仓计算
    position_query = Position.all()
    if account_id is not None:
        position_query = position_query.filter(account_id=account_id)
    positions = await position_query

    # 计算持仓成本
    total_cost = sum([pos.quantity * pos.avg_cost for pos in positions])

    # 计算当前市值
    total_market_value = 0
    for pos in positions:
        price_data = await stock_price_service.get_latest_price(pos.stock_code)
        if price_data:
            total_market_value += pos.quantity * price_data.price

    total_profit = total_market_value - total_cost
    return_rate = total_profit / total_cost if total_cost > 0 else 0

    # 从交易记录计算
    trade_query = Trade.all()
    if account_id is not None:
        trade_query = trade_query.filter(account_id=account_id)
    if start_date:
        trade_query = trade_query.filter(trade_date__gte=start_date)
    if end_date:
        trade_query = trade_query.filter(trade_date__lte=end_date)

    trades = await trade_query
    winning_trades = len([t for t in trades if t.profit and t.profit > 0])
    total_trades = len([t for t in trades if t.trade_type == "sell"])
    win_rate = winning_trades / total_trades if total_trades > 0 else 0

    return StatisticsResponse(
        total_profit=total_profit,
        total_cost=total_cost,
        total_market_value=total_market_value,
        return_rate=return_rate,
        win_rate=win_rate,
        total_trades=total_trades,
        winning_trades=winning_trades,
        losing_trades=total_trades - winning_trades,
        avg_holding_days=0
    )


async def get_summary() -> SummaryResponse:
    accounts = await Account.all()
    positions = await Position.all()
    trades = await Trade.all()

    total_cost = sum([pos.quantity * pos.avg_cost for pos in positions])

    total_market_value = 0
    for pos in positions:
        price_data = await stock_price_service.get_latest_price(pos.stock_code)
        if price_data:
            total_market_value += pos.quantity * price_data.price

    total_profit = total_market_value - total_cost

    return SummaryResponse(
        total_accounts=len(accounts),
        total_positions=len(positions),
        total_trades=len(trades),
        total_cost=total_cost,
        total_profit=total_profit
    )

File: stock_platform/java/backend/java/src/main/java/com/stockplatform/service/StatisticsService.java
package com.stockplatform.service;

import com.my.mybatis.flex.core.condition.QueryCondition;
import com.stockplatform.dto.StatisticsDTO;
import com.stockplatform.dto.SummaryDTO;
import com.stockplatform.entity.Account;
import com.stockplatform.entity.Position;
import com.stockplatform.entity.StockPrice;
import com.stockplatform.entity.Trade;
import com.stockplatform.mapper.AccountMapper;
import com.stockplatform.mapper.PositionMapper;
import com.stockplatform.mapper.StockPriceMapper;
import com.stockplatform.mapper.TradeMapper;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.List;

@Service
public class StatisticsService {

    private final PositionMapper positionMapper;
    private final TradeMapper tradeMapper;
    private final StockPriceMapper stockPriceMapper;
    private final AccountMapper accountMapper;

    public StatisticsService(PositionMapper positionMapper, TradeMapper tradeMapper,
                              StockPriceMapper stockPriceMapper, AccountMapper accountMapper) {
        this.positionMapper = positionMapper;
        this.tradeMapper = tradeMapper;
        this.stockPriceMapper = stockPriceMapper;
        this.accountMapper = accountMapper;
    }

    public StatisticsDTO getStatistics(Integer accountId, LocalDateTime startDate, LocalDateTime endDate) {
        QueryCondition positionCondition = QueryCondition.create();
        if (accountId != null) {
            positionCondition.where("account_id = ?", accountId);
        }
        List<Position> positions = positionMapper.selectAllByCondition(positionCondition);

        double totalCost = 0;
        double totalMarketValue = 0;
        for (Position pos : positions) {
            totalCost += pos.getQuantity() * pos.getAvgCost();
            StockPrice price = getLatestPrice(pos.getStockCode());
            if (price != null) {
                totalMarketValue += pos.getQuantity() * price.getPrice();
            }
        }

        double totalProfit = totalMarketValue - totalCost;
        double returnRate = totalCost > 0 ? totalProfit / totalCost : 0;

        QueryCondition tradeCondition = QueryCondition.create();
        if (accountId != null) {
            tradeCondition.where("account_id = ?", accountId);
        }
        if (startDate != null) {
            tradeCondition.where("trade_date >= ?", startDate);
        }
        if (endDate != null) {
            tradeCondition.where("trade_date <= ?", endDate);
        }
        List<Trade> trades = tradeMapper.selectAllByCondition(tradeCondition);

        int winningTrades = 0;
        int sellTrades = 0;
        for (Trade trade : trades) {
            if ("sell".equalsIgnoreCase(trade.getTradeType())) {
                sellTrades++;
                if (trade.getProfit() != null && trade.getProfit() > 0) {
                    winningTrades++;
                }
            }
        }
        double winRate = sellTrades > 0 ? (double) winningTrades / sellTrades : 0;

        StatisticsDTO dto = new StatisticsDTO();
        dto.setTotalProfit(totalProfit);
        dto.setTotalCost(totalCost);
        dto.setTotalMarketValue(totalMarketValue);
        dto.setReturnRate(returnRate);
        dto.setWinRate(winRate);
        dto.setTotalTrades(sellTrades);
        dto.setWinningTrades(winningTrades);
        dto.setLosingTrades(sellTrades - winningTrades);
        dto.setAvgHoldingDays(0.0);
        return dto;
    }

    public SummaryDTO getSummary() {
        List<Account> accounts = accountMapper.selectAll();
        List<Position> positions = positionMapper.selectAll();
        List<Trade> trades = tradeMapper.selectAll();

        double totalCost = 0;
        double totalMarketValue = 0;
        for (Position pos : positions) {
            totalCost += pos.getQuantity() * pos.getAvgCost();
            StockPrice price = getLatestPrice(pos.getStockCode());
            if (price != null) {
                totalMarketValue += pos.getQuantity() * price.getPrice();
            }
        }

        double totalProfit = totalMarketValue - totalCost;

        SummaryDTO dto = new SummaryDTO();
        dto.setTotalAccounts(accounts.size());
        dto.setTotalPositions(positions.size());
        dto.setTotalTrades(trades.size());
        dto.setTotalCost(totalCost);
        dto.setTotalProfit(totalProfit);
        return dto;
    }

    private StockPrice getLatestPrice(String stockCode) {
        QueryCondition condition = QueryCondition.create()
                .where("stock_code = ?", stockCode)
                .orderBy("trade_date", true);
        List<StockPrice> prices = stockPriceMapper.selectAllByCondition(condition);
        return prices.isEmpty() ? null : prices.get(0);
    }
}